[docs] Update Push notifications and expo-notifications docs (#20270)
expo/expo: docs/pages/push-notifications/push-notifications-setup.mdx

---
title: Push notifications setup
sidebar_title: Setup
---

import { Tab, Tabs } from '~/ui/components/Tabs';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';
import ImageSpotlight from '~/components/plugins/ImageSpotlight';
import { BoxLink } from '~/ui/components/BoxLink';

To use Expo's push notification service, you need to configure your app to install a set of libraries, add functions to handle notifications and configure credentials for Android and iOS. After going through these steps, you'll be able to test notification sending and receiving notifications on a device.

> To get the client-side ready for push notifications, the following is required:

Reviewer: ```suggestion
To get the client-side ready for push notifications, the following things are required:
```

- The user's permission to send them push notifications

> - The user's [`ExpoPushToken`](/versions/latest/sdk/notifications/#expopushtoken) &mdash; if push notifications are mail, then the `ExpoPushToken` is the user's address.

Reviewer: ```suggestion
- The user's [`ExpoPushToken`](/versions/latest/sdk/notifications/#expopushtoken).
```
We can keep this extra copy if you'd like, but I don't think it's totally needed.

> You'll need to use [`expo-notifications`](/versions/latest/sdk/notifications) library to request for user's permission and fetch the `ExpoPushToken`. You'll also use [`expo-device`](/versions/latest/sdk/device) to check whether the app is running on a physical device. The `expo-notifications` library is not supported on an Android Emulator or an iOS simulator.

Reviewer: ```suggestion
You'll need to use the [`expo-notifications`](/versions/latest/sdk/notifications) library to request a user's permission and to fetch the `ExpoPushToken`. You'll also use [`expo-device`](/versions/latest/sdk/device) to check whether the app is running on a physical device. The `expo-notifications` library is not supported on an Android Emulator or an iOS simulator.
```
I think that this information would be much better next to the step where they are installed below. So instead of summarizing how something works here, then having a step below that installs it, we should (1) ask the developer to install the libraries and (2) explain that they won't work on simulators/emulators.

Author: Initially, I had the explanation after installation. I don't exactly remember why I changed it 😄 
I've now added the explanation back in the installation step.

  style={{ maxWidth: 1200 }}
/>

Make sure that the development server for your project is running. Then, open the project in the Expo Go app and after the `ExpoPushToken` is generated, enter the value in the Expo push notifications tool with other details (for example, a message title and body) and click on the **Send a Notification** button.

After sending the notification from the tool, you should see the notification on your device. Below is an example of an Android device receiving a push notification.

<ImageSpotlight
  alt="An Android device receiving a push notification."
  src="/static/images/notifications/notification-on-android.png"
  style={{ maxWidth: 360 }}
/>

</Step>

<Step label="4">

## Get Credentials for development builds

For Android and iOS, there are different requirements to set up your credentials.

### Android

For Android, you need to configure **Firebase Cloud Messaging (FCM)** to get your credentials and set up your Expo project. It is required for all Android apps using Expo SDK (unless you are testing your project in Expo Go).

> **warning** FCM is not currently available for `expo-notifications` on iOS.

> #### Setup client

Reviewer: ```suggestion
#### Setting up FCM
```

1. To create a Firebase project, go to the [Firebase console](https://console.firebase.google.com/) and click on **Add project**.

2. In the console, click the setting icon next to **Project overview** and open **Project settings**. Then, under **Your apps**, click the Android icon to open **Add Firebase to your Android app** and follow the steps. **Make sure that the Android package name you enter is the same as the value of `android.package` from your app.json.**

3. After registering the app, download the **google-services.json** file and place it in your project's root directory.

   > The **google-services.json** file contains unique and non-secret identifiers of your Firebase project. For more information, see [Understand Firebase Projects](https://firebase.google.com/docs/projects/learn-more#config-files-objects).

> 4. In the **app.json**, add an `android.googleServicesFile` field with the relative path to the downloaded **google-services.json** file. If place it in the root directory, the path is:

Reviewer: ```suggestion
4. In **app.json**, add an `android.googleServicesFile` field with the relative path to the downloaded **google-services.json** file. If you placed it in the root directory, the path is:
```

5. Under **Server Credentials** > **FCM Server Key**, click **Add a FCM Server Key** > **Google Cloud Messaging Token** and add the **Server key** from **step 3**.

   > **warning** Having Service Credentials in both Legacy and non-legacy Application Identifiers can prevent push notifications from working on Android devices. If you have a Legacy Application Identifier, you should remove all of its Service Credentials.

### iOS

> **warning** A paid Apple Developer Account is required to generate credentials.

For iOS, make sure you have [registered your iOS device](/development/create-development-builds/#on-a-device) on which you want to test before running the `eas build` command for the first time.

If you create a development build for the first time, you'll be asked to enable push notifications. Answer yes to the following questions when prompted:

- Setup Push Notifications for your project
- Generating a new Apple Push Notifications service key

<br />

> > If you are using bare workflow and not EAS Build, you will need to run `eas credentials` manually.

Reviewer: ```suggestion
> If you are not using EAS Build, you will need to run `eas credentials` manually.
```
